pkg/broker/broker_test.go

```diff
@@ -55,3 +56,93 @@ func TestAddIdForPlan(t *testing.T) {
 	addIDForPlan(plans, "dh-sns-apb")
 	ft.AssertNotEqual(t, plans[0].ID, "", "plan id not updated")
 }
+
+// note this would likely move to a mocks package
+type mockJobStateDAO struct {
+	state  apb.JobState
+	states []apb.JobState
+	err    error
+	calls  map[string]int
+}
+
+func (mjs *mockJobStateDAO) GetState(instanceUUID, operation string) (apb.JobState, error) {
+	mjs.calls["GetState"]++
+	return mjs.state, mjs.err
+}
+func (mjs *mockJobStateDAO) SetState(id string, state apb.JobState) error {
+	mjs.calls["SetState"]++
+	return mjs.err
+}
+func (mjs *mockJobStateDAO) GetSvcInstJobsByState(instanceID string, reqState apb.State) ([]apb.JobState, error) {
+	mjs.calls["GetSvcInstJobsByState"]++
+	return mjs.states, mjs.err
+}
+
+func TestAnsibleBroker_LastOperation(t *testing.T) {
+	cases := []struct {
+		Name                 string
+		JobStateDAO          func() JobStateDAO
+		ExpectError          bool
+		InstanceID           uuid.UUID
+		LastOperationRequest *LastOperationRequest
+		Validate             func(t *testing.T, resp *LastOperationResponse)
+		ExpectedCalls        map[string]int
```

Comment:
I would want to simplify this a bit. Definitely the right idea but this test looks very painful to debug and understand in 30s. I prefer a bunch of small tests vs a test that loops through all cases.

Reply:
Interesting. I want to be clear that I have no problem in changing this to follow the patterns you feel are best for this project, however in the interest of discussion I will explain my reasons here. You may already know all the following, if so feel free to ignore.

These kind of tests are quite common in go projects, including the stdlib, the built in testing package has builtin support for running sub tests in this way. 

Using the ```t.Run ``` method runs each of these cases as their own test. This produces output as follows:

```
--- PASS: TestAnsibleBroker_LastOperation (0.00s)
    --- PASS: TestAnsibleBroker_LastOperation/test_get_last_operation_successful
```
I think this is quite clear and gives a nice view of what has been tested. Also if the sub test: ``` TestAnsibleBroker_LastOperation/test_get_last_operation_successful ``` fails, you can re run just that single sub test by using ```go test -run ^TestAnsibleBroker_LastOperation/test_get_last_operation_successful$``` to isolate it down, no matter how many cases may be present. This should allow you to debug quite quickly if an individual case is failing.
In my experience have a structure for your test case makes it quite simple not only to add a new case but also to understand the pieces involved in the test and what can be controlled.

As mentioned these are only my own thoughts and I will happily adjust.

Reply:
@maleck13 well you taught me something new today. It's an interesting approach for doing positive and negative testing on a single piece of code. The examples on the go blog were easier since they just had strings in their test data. I would not be opposed to it using table driven tests.